packages/airnode-utilities/src/evm/gas-prices/gas-oracle.ts

```diff
@@ -0,0 +1,296 @@
+
+export const calculateTimeout = (startTime: number, totalTimeout: number) => totalTimeout - (Date.now() - startTime);
+
+export const parsePriorityFee = ({ value, unit }: config.Amount) => ethers.utils.parseUnits(value.toString(), unit);
+
+export const multiplyGasPrice = (gasPrice: ethers.BigNumber, gasPriceMultiplier: number) =>
+  gasPrice.mul(ethers.BigNumber.from(Math.round(gasPriceMultiplier * 100))).div(ethers.BigNumber.from(100));
+
+// This will return the gasLimit with the gasTarget if getGasPrice is called from a place where the fulfillmentGasLimit
+// is available in the config.json and otherwise no gasLimit will be returned for use with other transactions
+// (e.g. withdrawal transactions and airnode-admin CLI/SDK)
+export const getGasLimit = (fulfillmentGasLimit?: number) => {
+  return fulfillmentGasLimit ? { gasLimit: ethers.BigNumber.from(fulfillmentGasLimit) } : {};
+};
+
+// Returns the gasTarget with gasLimit from fulfillmentGasLimit added
+export const getGasTargetWithGasLimit = (gasTarget: GasTarget, fulfillmentGasLimit?: number) => ({
+  ...gasTarget,
+  ...getGasLimit(fulfillmentGasLimit),
+});
+
+export const getPercentile = (percentile: number, array: ethers.BigNumber[]) => {
+  if (!array.length) return;
+  array.sort((a, b) => (a.gt(b) ? 1 : -1));
+
+  const index = Math.ceil(array.length * (percentile / 100)) - 1;
+  return array[index];
+};
+
+// Check whether a value's change exceeds the maxDeviationMultipiler limit
+// and returns false if it does and true otherwise.
+export const checkMaxDeviationLimit = (
+  value: ethers.BigNumber,
+  referenceValue: ethers.BigNumber,
+  maxDeviationMultiplier: number
+) => {
+  // Handle maximum two decimals for maxDeviationMultiplier
+  const maxDeviationMultiplierBN = ethers.BigNumber.from(Math.round(maxDeviationMultiplier * 100));
+
+  return (
+    // Check that the current value is not higher than the maxDeviationMultiplier allows
+    referenceValue.gt(value.mul(ethers.BigNumber.from(100)).div(maxDeviationMultiplierBN)) &&
+    // Check that the current value is not lower than the maxDeviationMultiplier allows
+    referenceValue.lt(value.mul(maxDeviationMultiplierBN).div(ethers.BigNumber.from(100)))
+  );
+};
+
+// Returns the constant gas price
+export const fetchConstantGasPrice = (constantGasPriceStrategy: config.ConstantGasPriceStrategy): LegacyGasTarget => ({
+  type: 0,
+  gasPrice: parsePriorityFee(constantGasPriceStrategy.gasPrice),
+});
+
+// Returns the provider gas price and applies the recommended multiplier
+export const fetchProviderRecommendedGasPrice = async (
+  provider: Provider,
+  gasOracleOptions: config.ProviderRecommendedGasPriceStrategy
+): Promise<LegacyGasTarget> => {
+  const { recommendedGasPriceMultiplier } = gasOracleOptions;
+
+  const goGasPrice = await go(() => provider.getGasPrice(), {
+    onAttemptError: (goError) => logger.warn(`Failed attempt to get gas price. Error: ${goError.error}.`),
```

Comment:
`onAttemptError` will have no effect (will never be called) if you have no retries. For just one function call the error is available as a return value.